### ISSUE TITLE
SSO Sign In for Client

### ISSUE THREAD

**Issue by lawrence-vo** [fixer-m/snowflake-db-net-client#28]
Is there anyway to authenticate via. SSO? I don't mind taking a crack at this, just need to know if it's been implemented or not.

Thanks!

**Comment by fixer-m** [comment]
Hi! No, unfortunately this library doesn't support this yet. 
You can create dedicated user in SF, allow auth with password for it and use it for this client.

**Comment by lawrence-vo** [comment]
@fixer-m would you like me to take a crack at SSO or would you prefer that be implemented at a later time?

**Comment by fixer-m** [comment]
@lawrence-vo Sure - you can try to implement it, this would be great actually. I have only demo account in SF, so I guess I can't login via SSO and thus can't properly try and test this auth scheme. I would start with analysing how SSO works in official connector, try to recreate it and simplify it if possible.

**Comment by lawrence-vo** [comment]
@fixer-m might take longer as expected since the official client also has issues with SSO.

**Comment by Tsanton** [comment]
@lawrence-vo can you link to the issue with the official client?

I just implemented a client credentials flow (OAuth) between DBT and Azure AD for Snowflake.
The project I'm working on requires OAuth-authentication in prod and I'm also utilising this client, therefore I too wouldn't mind taking a stab at as I would love to use this package.

@fixer-m do you have any design ideas for how to implement multiple authentication alternatives?

I'm thinking a `IAuthenticator` interface... ?

If we can agree on some common patterns then I'd have no problems with creating the AzureOauthAuthenication implementation!

**Comment by fixer-m** [comment]
@Tsanton Yes, I guess there should be `IAuthenticator` interface with multiple implementations. I don't know much about OAuth, so right now I can't suggest any design patterns (I would start from researching official connector code). 

I have a few requirements though: 
- Try to not break existing public contracts
- It should be easy to use. Use some common practices and standards
- It should be easy to add new auth implementations

**Comment by lawrence-vo** [comment]
@Tsanton , the [issue] is here (https://github.com/snowflakedb/snowflake-connector-net/issues/308)
